src/types.ts

```diff
@@ -63,26 +63,24 @@ export interface ProtocolWithReturn<Data, Return> {
  * Extendable by user.
  */
 export interface ProtocolMap {
-  // foo: { id: number, name: string }
+  __crx_bridge_stream_open__: ProtocolWithReturn<{ channel: string; streamId: string }, boolean>
+  __crx_bridge_stream_transfer__: { streamId: string; action: 'transfer' | 'close'; streamTransfer: JsonValue }
+  // foo: { id: number; name: string; }
   // bar: ProtocolWithReturn<string, number>
 }
 
 export type DataTypeKey = keyof ProtocolMap
 
+type ValidateJsonValue<T> = T extends JsonValue ? T : 'This value is not a serializable JSON value'
```

Comment:
It's not possible to restrict how one extends `ProtocolMap` but we can restrict what can be consumed from the `ProtocolMap` at the "call sites":
<img width="334" alt="Screenshot 2022-06-30 at 11 55 20" src="https://user-images.githubusercontent.com/9800850/176649345-889ff372-f370-406e-87cf-4775a3bc7db2.png">
<img width="1048" alt="Screenshot 2022-06-30 at 11 54 40" src="https://user-images.githubusercontent.com/9800850/176649351-9ac6e1b8-aa78-4b73-966a-e30913c23cbb.png">
<img width="983" alt="Screenshot 2022-06-30 at 11 54 21" src="https://user-images.githubusercontent.com/9800850/176649357-e2949b6c-1850-484e-a2f6-354cc8182243.png">